### ISSUE TITLE
Push Publishing Exclusion Filters

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#17806]
### Problem:

When content is pushed from one system to another, dotCMS tries to "intelligently" decide what content should be pushed as a dependency by consulting the push history.  Many times in a publishing environment your auth and prod are mostly in sync and you do not want to include any dependencies \(shallow push\), or only include a subset of dependencies and exclude others \(no users, no hosts, no workflow, no content types\).

Currently, as an example, hosts always seem to get pushed from authoring environments to production environments, or from one environment to another where they should normally be excluded as a dependency once the two environments have been synced.
That is not to say you should not be able to push the dependent objects, only that 9 times out of 10,  that they should not be sent as a dependency by default.

### Possible Solution

We should allow the creation of PP exclusion filters.  These filters can have a name, and they should specify what CMS objects should be excluded when pushing.  When creating a new PP environment, an admin can specify which filter should be used by default and when business users PP content to those environments, the environment filters should be respected with regards to what is included/excluded in any bundle as dependencies.   The exclusions happens at bundle time.  If the user is a CMS admin, perhaps they are allowed to select from the list of available filters when firing/scheduling a PP to allow them to override the default behavior.

Exclusion Filters will be stored as YAML in the `./assets/server/publishing-filters` folder

```
key:                      (this would be the filename)
title:                    (arbitrary title for the exclude filter)
default:                  true|false
roles:                     [list of roles key]
filters:
       excludeDependencyQuery:   ( lucene query that dependent content is checked against to exclude)
       excludeQuery:             (lucene query that added content is checked against to exclude)
       excludeDependencyClasses: [ list of classes that should never be pushed as dependencies]
       excludeClasses:           [ list of classes that should never be pushed]
       forcePush:              true|false
       dependencies:             true|false, defaults to true
       relationships:            true|false, defaults to true
```

Initial Filters:
  - Shallow Push \(only pushes the object with NO dependencies\)
  - Content Only \(only pushes content and content relationships
  - Web Content Only \(folders, all content, menu links, language variables,  files, vanity urls, rules on pages\).  This does not push Host/Workflow/Content Type\)
  - Intelligent \(current state, respecting push history\)
  - Force Push \(as it does now\)

**Comment by jcastro-dotcms** [comment]
This is great! Many customers have asked about this several times. And also, customers have asked that if we had add some setting **not** overwrite Host Aliases when pushing Sites, it'd be very nice. For example: [100159](https://dotcms.zendesk.com/agent/tickets/100159)

**Comment by erickgonzalez** [comment]
Please ignore previous commits. This issue has been converted to Epic and subdivided into smaller issues.

**Comment by erickgonzalez** [comment]
Some draft doc https://docs.google.com/document/d/1O0sFT7WSj_FJ_Q4c3rHSZbOiA465YVBaFcsFtIlUdPo/edit?usp=sharing

**Comment by bryanboza** [comment]
After test this new functionality we some work in those cases: 

- We need to hide this functionality from the push remove option,  this because we are not respecting the filters in the remove process

- `Content and Relationships` filter is not pushing relationships: 
_Create a new content type and relate to itself
Add two new pieces of content and relate those ones 
Push using that filter, we are sending just the selected item_

-`Content, Assets and pages` filter is not pushing relationships 

- Need to improve the error message when you try to send just a piece of content using the filter `only selected content` and the content type does not exist on receiver 
SC: https://www.screencast.com/t/QX0BVfQWtE

- When you push using the `only selected content` filter we allow send required values in null
SC: https://www.screencast.com/t/5rJm3eQhaI

**Comment by erickgonzalez** [comment]
> We need to hide this functionality from the push remove option, this because we are not respecting the filters in the remove process

Yes, we need to hide it from the **REMOVE** option in the PP modal. And also doc that we using **PUSH-REMOVE** the filters will be applied for the **PUSH** but no the **REMOVE**. This because when we **REMOVE** we do not search for dependencies of the content (this is how PP has always worked).

> Content and Relationships filter is not pushing relationships:
Create a new content type and relate to itself
Add two new pieces of content and relate those ones
Push using that filter, we are sending just the selected item

Able to reproduce even with the `Everything and Dependencies` Filter, looking at it because could be an issue related to relationships. https://github.com/dotCMS/core/issues/18775

> Need to improve the error message when the content type does not exist on the receiver

In the log shows the error that the content type does not exist, but we can improve this message.

> When you push using the only selected content filter we allow send required values in null

We are discussing how the `ONLY SELECTED CONTENT FILTER` should be over here https://github.com/dotcms/core/issues/18742 

> The path where are located the yml files of filters maybe should be configurable

Could be an enhancement for a future release.

**Comment by fmontes** [comment]
> We need to hide this functionality from the push remove option, this because we are not respecting the filters in the remove process
PR: https://github.com/dotCMS/core-web/pull/1364

**Comment by fmontes** [comment]
Fixing 1 and 2 that are blockers, all the others needs discussion and priority.

**Comment by fmontes** [comment]
## Fixed:

> We need to hide this functionality from the push remove option, this because we are not respecting the filters in the remove process

PR: dotCMS/core-web#1364

> Content and Relationships filter is not pushing relationships:
Create a new content type and relate to itself
Add two new pieces of content and relate those ones
Push using that filter, we are sending just the selected item

https://github.com/dotCMS/core/issues/18775

**Comment by bryanboza** [comment]
Fixed, all the reported cases was moved to new cards. Tested on release-5.3.3 // Postgres // FF

**Comment by bryanboza** [comment]
After clean the value of this filed this can cause confusion, maybe we need to add some text like "filters not apply for remove publish" to be clear that functionality is not available in the remove process  or totally hide the field. 
![image](https://user-images.githubusercontent.com/2641437/85909575-de152200-b7d7-11ea-9d12-f039d90dca1e.png)

**Comment by fmontes** [comment]
Asked @wezell agreed on not doing this last comment @bryanboza

**Comment by bryanboza** [comment]
Fixed, tested on release-5.3.3 // Postgres // FF
